internal/releasesources/release_sources.go

```diff
@@ -0,0 +1,478 @@
+	resultReleases := []*common.ReleaseImage{}
+	releaseImageReferences := map[string]bool{}
+
+	for _, releaseImage := range releases {
+		if releaseImageReferences[*releaseImage.URL] {
+			continue
+		}
+
+		releaseImageReferences[*releaseImage.URL] = true
+		resultReleases = append(resultReleases, releaseImage)
+	}
+
+	return resultReleases
+}
+
+func (h *ReleaseSourcesHandler) removeReleaseImageDuplicates(releases []*common.ReleaseImage) []*common.ReleaseImage {
+	releases = h.removeReleaseImageDuplicatesFromDifferentChannels(releases)
+	releases = h.removeReleaseImageDuplicateRecords(releases)
+	return releases
+}
+
+// getReleasesWithoutDuplicates removes duplicate release images from different channels
+func (h *ReleaseSourcesHandler) removeReleaseImageDuplicatesFromDifferentChannels(releases []*common.ReleaseImage) []*common.ReleaseImage {
+	resultReleases := []*common.ReleaseImage{}
+	stableReleasesReferences := map[string]bool{}
+
+	for _, release := range releases {
+		if *release.Channel == OpenshiftReleaseChannelStable {
+			stableReleasesReferences[*release.URL] = true
+		}
+	}
+
+	// Add only releases which does not have a stable duplicate
+	for _, release := range releases {
+		if *release.Channel != OpenshiftReleaseChannelStable && stableReleasesReferences[*release.URL] {
+			continue
+		}
+
+		resultReleases = append(resultReleases, release)
+	}
+
+	return resultReleases
+}
+
+func (h *ReleaseSourcesHandler) getDynamicReleaseImages() ([]*common.ReleaseImage, error) {
+	// OCP releases API needs amd64, arm64 instead of x86_64, aarch64 respectively
+	cpuArchMapToAPIArch := map[string]string{
+		common.X86CPUArchitecture:     common.AMD64CPUArchitecture,
+		common.AARCH64CPUArchitecture: common.ARM64CPUArchitecture,
+	}
+
+	// switch back
+	cpuArchMapFromAPIArch := map[string]string{
+		common.AMD64CPUArchitecture: common.X86CPUArchitecture,
+		common.ARM64CPUArchitecture: common.AARCH64CPUArchitecture,
+	}
+
+	releases := []*common.ReleaseImage{}
+	// Aggregate slice of releases from releases API
+	for _, releaseSource := range h.releaseSources {
+		openshiftVersion := *releaseSource.OpenshiftVersion
+		for _, upgradeChannel := range releaseSource.UpgradeChannels {
+			cpuArchitecture := *upgradeChannel.CPUArchitecture
+			for _, channel := range upgradeChannel.Channels {
+				apiCpuArchitecture, shouldSwitch := cpuArchMapToAPIArch[cpuArchitecture]
+				if shouldSwitch {
+					cpuArchitecture = apiCpuArchitecture
+				}
+
+				graph, err := h.releasesClient.GetReleases(channel, openshiftVersion, cpuArchitecture)
+				if err != nil {
+					h.log.WithError(err).Warnf("failed to get releases from api with: %s-%s, %s", channel, openshiftVersion, cpuArchitecture)
+					continue
+				}
+
+				if shouldSwitch {
+					if arch, ok := cpuArchMapFromAPIArch[cpuArchitecture]; ok {
+						cpuArchitecture = arch
+					}
+				}
+
+				for _, node := range graph.Nodes {
+					if strings.HasPrefix(node.Version, openshiftVersion) {
```

Comment:
will not 4.11 match 4.1 in this case?

Reply:
Yes basically the results are only for the specified version or adjacent ones e.g. "4.13", "4.14", "4.15" but I agree I should change it to avoid this matter, thank you